lib/maths/unittest/CBoostedTreeTest.cc

```diff
@@ -1432,22 +1432,22 @@ BOOST_AUTO_TEST_CASE(testProgressMonitoring) {
                 // We don't do feature selection (we have enough data to use all of them).
             } else if (task.s_Name == maths::CBoostedTreeFactory::COARSE_PARAMETER_SEARCH) {
                 // We don't have accurate upfront estimate of the number of steps so we
-                // only get progress up to 80%. In non-test code we always pass 100% when
-                // the task is complete.
-                BOOST_REQUIRE_EQUAL("[0, 10, 20, 30, 40, 50, 60, 70, 80]",
-                                    core::CContainerPrinter::print(task.s_TenPercentProgressPoints));
+                // only get progress up to 80% or 90% depending on the compiler and
+                // platform. In non-test code we always pass 100% when the task is complete.
+                if (task.s_TenPercentProgressPoints.size() != 10 ||
+                    task.s_TenPercentProgressPoints.front() != 0 ||
+                    task.s_TenPercentProgressPoints.back() != 90) {
+                    BOOST_REQUIRE_EQUAL("[0, 10, 20, 30, 40, 50, 60, 70, 80]",
+                                        core::CContainerPrinter::print(task.s_TenPercentProgressPoints));
+                }
             } else if (task.s_Name == maths::CBoostedTreeFactory::FINE_TUNING_PARAMETERS) {
                 BOOST_REQUIRE_EQUAL("[0, 10, 20, 30, 40, 50, 60, 70, 80, 90]",
                                     core::CContainerPrinter::print(task.s_TenPercentProgressPoints));
             } else if (task.s_Name == maths::CBoostedTreeFactory::FINAL_TRAINING) {
-                // Progress might be 90% or 100% depending on whether the final
-                // progress update registered
-                if (task.s_TenPercentProgressPoints.size() != 11 ||
-                    task.s_TenPercentProgressPoints.front() != 0 ||
-                    task.s_TenPercentProgressPoints.back() != 100) {
-                    BOOST_REQUIRE_EQUAL("[0, 10, 20, 30, 40, 50, 60, 70, 80, 90]",
-                                        core::CContainerPrinter::print(task.s_TenPercentProgressPoints));
-                }
+                // Just assert that training started as there is no guarantee on how far it's got
+                BOOST_TEST_REQUIRE(task.s_TenPercentProgressPoints.empty() == false);
+                BOOST_TEST_REQUIRE("[0]" != core::CContainerPrinter::print(
+                                                task.s_TenPercentProgressPoints));
```

Comment:
This is interesting. Did you see cases where progress where we got [0, 10]? For final training we know how many trees we plan to use so it is surprising that this is very variable.

Reply:
No, I only ever saw 70 to 100.  But if this is at the mercy of thread scheduling on the test machine then all we can safely say is that it's started.  Asserting 70 to 100 will probably be OK for a few years and then we'll start testing on a machine with different timing characteristics and we'll get a 50 or 60.

Reply:
This test isn't actually affected by threading. The reason it can finish with less than 100% progress is we estimate the number of steps upfront and this estimate can be too large. That said this need not hold this PR up further. I'll maybe revisit these checks in a different change.

Reply:
Oh, sorry, I misunderstood what the problem was.